### PR TITLE
flutter: add darwin support

### LINKED ISSUE #145701: Cannot install flutter on osX
### Describe the bug
I'd like to install flutter on osX

### Steps To Reproduce
1. nix-env -iA nixpkgs.flutter

### Expected behavior
Install flutter

### Screenshots
```
warning: name collision in input Nix expressions, skipping '/Users/primet/.nix-defexpr/channels_root/home-manager'
installing 'flutter-2.5.1'
error: unsupported version/system: 2.14.2/x86_64-darwin
(use '--show-trace' to show detailed location information)
```

### Notify maintainers
@babariviere @ericdallo @thiagokokada 

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-darwin"`
 - host os: `Darwin 20.6.0, macOS 10.16`
 - multi-user?: `no`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.3.15`
 - channels(primet): `"home-manager-21.05"`
 - channels(root): `"darwin, home-manager-21.05, nixpkgs-21.11pre320922.ee084c02040"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixpkgs`
```

### PR DESCRIPTION
**Pull request by devholic**
###### Description of changes

> Based on https://github.com/NixOS/nixpkgs/pull/196133

Adds `x86_64`/`aarch64`-darwin support for flutter

---

EDIT: Closes #145701

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/compilers/flutter/default.nix [hunk 1]
```diff
@@ -17,12 +17,34 @@ let
...
+    src = rec {
+      x86_64-linux = fetchurl {
+        url = "https://storage.googleapis.com/flutter_infra_release/releases/stable/linux/flutter_linux_${version}-stable.tar.xz";
+        sha256 = flutterHash.x86_64-linux;
+      };
+      aarch64-linux = fetchurl {
+        url = "https://storage.googleapis.com/flutter_infra_release/releases/stable/linux/flutter_linux_${version}-stable.tar.xz";
+        sha256 = flutterHash.aarch64-linux;
+      };
+      x86_64-darwin = fetchzip {
```
Comment (wegank) [inline pkgs/development/compilers/flutter/default.nix:39]: Can `fetchurl` be used here?

Comment (devholic) [reply pkgs/development/compilers/flutter/default.nix:39]: I tried to replace `fetchzip` with `fetchurl`, but the following error came out 🤔 

```
unpacking sources
unpacking source archive /nix/store/njydk0l4j8xjlah76a1261dbsrmn8drz-flutter_macos_3.3.8-stable.zip
do not know how to unpack source archive /nix/store/njydk0l4j8xjlah76a1261dbsrmn8drz-flutter_macos_3.3.8-stable.zip
error: builder for '/nix/store/qscx3g5riw142lsj75adiv3pimliprp1-flutter-3.3.8-unwrapped.drv' failed with exit code 1;
       last 3 log lines:
       > unpacking sources
       > unpacking source archive /nix/store/njydk0l4j8xjlah76a1261dbsrmn8drz-flutter_macos_3.3.8-stable.zip
       > do not know how to unpack source archive /nix/store/njydk0l4j8xjlah76a1261dbsrmn8drz-flutter_macos_3.3.8-stable.zip
```

Comment (wegank) [reply pkgs/development/compilers/flutter/default.nix:39]: Oh, I guess you need to add `unzip` to `nativeBuildInputs`.

Comment (devholic) [reply pkgs/development/compilers/flutter/default.nix]: Seems working well 👍 Thanks @wegank !

File: pkgs/development/compilers/flutter/patches/flutter3/move-cache-darwin.patch [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+         fileSystem.path.join('/out', 'host_debug_unopt', 'impellerc'),
+diff --git a/packages/flutter_tools/bin/podhelper.rb b/packages/flutter_tools/bin/podhelper.rb
+index 54238860d5..249ea8c087 100644
+--- a/packages/flutter_tools/bin/podhelper.rb
++++ b/packages/flutter_tools/bin/podhelper.rb
+@@ -38,7 +38,7 @@ def flutter_additional_ios_build_settings(target)
+ 
+   # This podhelper script is at $FLUTTER_ROOT/packages/flutter_tools/bin.
+   # Add search paths from $FLUTTER_ROOT/bin/cache/artifacts/engine.
+-  artifacts_dir = File.join('..', '..', '..', '..', 'bin', 'cache', 'artifacts', 'engine')
```
[the remark this answers is not in the document]
Comment (devholic) [reply pkgs/development/compilers/flutter/patches/flutter3/move-cache-darwin.patch:200]: cc @tanshihaj

**Comment by devholic** [comment]
@wegank Could you take a look?

**Comment by wegank** [comment]
The actual difference between the two patches is minor:

```diff
124c124,126
< @@ -588,7 +590,7 @@ class CachedArtifacts implements Artifacts {
---
> @@ -586,14 +588,10 @@ class CachedArtifacts implements Artifacts {
>          // For script snapshots any gen_snapshot binary will do. Returning gen_snapshot for
>          // android_arm in profile mode because it is available on all supported host platforms.
126,127c128,129
<        case Artifact.frontendServerSnapshotForEngineDartSdk:
<          return _fileSystem.path.join(
---
> -      case Artifact.frontendServerSnapshotForEngineDartSdk:
> -        return _fileSystem.path.join(
129,131c131,132
< +          _dartSdkPath(_fileSystem), 'bin', 'snapshots',
<            _artifactToFileName(artifact),
<          );
---
> -          _artifactToFileName(artifact),
> -        );
133c134,140
< @@ -776,7 +778,7 @@ class CachedLocalEngineArtifacts implements LocalEngineArtifacts {
---
>        case Artifact.vmSnapshotData:
>        case Artifact.isolateSnapshotData:
> +      case Artifact.frontendServerSnapshotForEngineDartSdk:
>        case Artifact.icuData:
>          final String engineArtifactsPath = _cache.getArtifactDirectory('engine').path;
>          final String platformDirName = _enginePlatformDirectoryName(platform);
> @@ -776,7 +774,7 @@ class CachedLocalEngineArtifacts implements LocalEngineArtifacts {
```

So not unifying them is a blocker for me.

**Comment by flokli** [comment]
> The actual difference between the two platform-specific patches is minor:
> […]

In nixpkgs, we usually apply patches unconditionally for all architectures, so e.g. a version bump, and resulting breakage of the patch gets spotted easily. Is it possible to merge these two patches, so they work on both architectures?

**Comment by n8henrie** [comment]
Just made [this PR](https://github.com/devholic/nixpkgs/pull/1/files) that merges the patches and builds on both aarch64-darwin and x86_64-linux.

Have never used flutter (was just tinkering this morning), so I'd appreciate if someone can confirm that it *works* (not just builds), but `./result/bin/flutter --version` seems to work on both machines.

Note to self: learn how to use `nixpkgs-review`.

**Comment by n8henrie** [comment]
Gah, NM. Needs a pretty major rebase, I probably shouldn't dive in until I actually know how to use it a little bit so I can make sure it seems to work.

**Comment by tlvenn** [comment]
@n8henrie would be very happy to help test your PR if you get the chance / time to do the rebase. Would be super helpful to finally get nix flutter package for darwin.

@devholic thanks for the initial PR, hopefully we can get this merged.

**Comment by n8henrie** [comment]
Unfortunately I've abandoned it for now, the rebase was far from trivial, and I have no need for flutter (was just exploring a hello world).

Sorry, I'm sure I'll look at this again eventually as I am still curious, but no immediate plans to continue.

**Comment by devholic** [comment]
@tlvenn I'm looking into it, but it might take some time since it changed a lot 😢

**Comment by devholic** [comment]
@tlvenn I did basic tests, and I think it's ready 🚀

**Review by wegank — APPROVED** [review]

**Review by thiagokokada — APPROVED** [review]

File: pkgs/development/compilers/flutter/patches/flutter310/move-cache.patch [hunk 3]
```diff
@@ -0,0 +1,152 @@
+diff --git a/packages/flutter_tools/bin/podhelper.rb b/packages/flutter_tools/bin/podhelper.rb
+index 01d0bf4e21..cb00a44674 100644
+--- a/packages/flutter_tools/bin/podhelper.rb
++++ b/packages/flutter_tools/bin/podhelper.rb
+@@ -39,7 +39,7 @@ def flutter_additional_ios_build_settings(target)
+
+   # This podhelper script is at $FLUTTER_ROOT/packages/flutter_tools/bin.
+   # Add search paths from $FLUTTER_ROOT/bin/cache/artifacts/engine.
+-  artifacts_dir = File.join('..', '..', '..', '..', 'bin', 'cache', 'artifacts', 'engine')
++  artifacts_dir = File.join(ENV['HOME'], '.cache/flutter/artifacts/engine')
```
Comment (FlafyDev) [inline pkgs/development/compilers/flutter/patches/flutter310/move-cache.patch:10]: Does darwin use .cache directory in HOME for artifacts? Currently Flutter in nixpkgs only uses .cache for the lockfile and everything else is readonly in `<flutter>/bin/cache`

File: pkgs/development/compilers/flutter/patches/flutter310/move-cache.patch [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+diff --git a/packages/flutter_tools/lib/src/artifacts.dart b/packages/flutter_tools/lib/src/artifacts.dart
+index a874b3c703..ba5ddd6600 100644
+--- a/packages/flutter_tools/lib/src/artifacts.dart
++++ b/packages/flutter_tools/lib/src/artifacts.dart
+@@ -643,7 +643,7 @@ class CachedArtifacts implements Artifacts {
+       case Artifact.dart2wasmSnapshot:
+       case Artifact.frontendServerSnapshotForEngineDartSdk:
+         return _fileSystem.path.join(
+-          _dartSdkPath(_cache), 'bin', 'snapshots',
++          _dartSdkPath(_fileSystem), 'bin', 'snapshots',
```
Comment (FlafyDev) [inline pkgs/development/compilers/flutter/patches/flutter310/move-cache.patch:32]: why is it changed from _cache to _fileSystem in several places in the patch? And could the reason be documented somewhere? maybe in the git commit message

File: pkgs/development/compilers/flutter/patches/flutter310/move-cache.patch [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
++    const Platform platform = LocalPlatform();
++    if (platform.environment.containsKey('FLUTTER_CACHE_DIR')) {
++      return _fileSystem.directory(platform.environment['FLUTTER_CACHE_DIR']);
++    }
++
+     if (_rootOverride != null) {
+       return _fileSystem.directory(_fileSystem.path.join(_rootOverride!.path, 'bin', 'cache'));
+     } else {
+-      return _fileSystem.directory(_fileSystem.path.join(flutterRoot!, 'bin', 'cache'));
++      return _fileSystem.directory(_fileSystem.path.join(globals.fsUtils.homeDirPath!, '.cache', 'flutter'));
```
Comment (FlafyDev) [inline pkgs/development/compilers/flutter/patches/flutter310/move-cache.patch:122]: What's the point of changing it? Wouldn't FLUTTER_CACHE_DIR will be set anyway?

**Review by FlafyDev — CHANGES_REQUESTED** [review]
I can't test on darwin, but I still left some questions..

Everything seems to still work on Linux:

Result of `nixpkgs-review pr 210067` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>13 packages built:</summary>
  <ul>
    <li>firmware-updater</li>
    <li>firmware-updater.debug</li>
    <li>fluffychat</li>
    <li>fluffychat.debug</li>
    <li>flutter</li>
    <li>flutter-unwrapped (flutterPackages.stable)</li>
    <li>flutter-unwrapped.cache (flutterPackages.stable.cache)</li>
    <li>flutter37</li>
    <li>flutter37-unwrapped (flutterPackages.v37)</li>
    <li>flutter37-unwrapped.cache (flutterPackages.v37.cache)</li>
    <li>hover</li>
    <li>yubioath-flutter</li>
    <li>yubioath-flutter.debug</li>
  </ul>
</details>

File: pkgs/development/compilers/flutter/wrapper.nix [hunk 6]
```diff
@@ -158,7 +160,10 @@ let
 in
 (callPackage ./sdk-symlink.nix { }) (runCommandLocal "flutter-wrapped"
 {
-  nativeBuildInputs = [ makeWrapper ];
+  nativeBuildInputs = [
+    makeWrapper
+    unzip
```
Comment (hacker1024) [inline pkgs/development/compilers/flutter/wrapper.nix:165]: Why `unzip`?

File: pkgs/development/compilers/flutter/patches/flutter3/move-cache.patch
Comment (hacker1024) [inline pkgs/development/compilers/flutter/patches/flutter3/move-cache.patch:1]: Why don't we have this patch for the latest version anymore? Or am I missing something?

**Review by hacker1024 — CHANGES_REQUESTED** [review]

File: pkgs/development/compilers/flutter/patches/flutter310/move-cache.patch (hunk 3 above)
Comment (devholic) [reply pkgs/development/compilers/flutter/patches/flutter310/move-cache.patch:10]: This PR is based on https://github.com/NixOS/nixpkgs/pull/196133, so not sure about this 🤔

File: pkgs/development/compilers/flutter/flutter.nix [hunk 7]
```diff
@@ -18,6 +20,8 @@ let
       outputs = [ "out" "cache" ];
 
       buildInputs = [ git ];
+      nativeBuildInputs = [ unzip ]
+        ++ lib.optionals stdenv.hostPlatform.isDarwin [ darwin.DarwinTools ];
```
[the remark this answers is not in the document]
Comment (devholic) [reply pkgs/development/compilers/flutter/flutter.nix:24]: There are mac specific dependencies in Flutter SDK (e.g. `sw_ver`), so I put DarwinTools to the nativeBuildInputs.

File: pkgs/development/compilers/flutter/patches/flutter3/move-cache.patch
Comment (devholic) [reply pkgs/development/compilers/flutter/patches/flutter3/move-cache.patch:1]: This patch is moved to `flutter310` directory ([ref](https://github.com/NixOS/nixpkgs/blob/441b1944a35a205627904be3c4456d054be67b6c/pkgs/development/compilers/flutter/patches/flutter310/move-cache.patch))

File: pkgs/development/compilers/flutter/wrapper.nix (hunk 6 above)
Comment (devholic) [reply pkgs/development/compilers/flutter/wrapper.nix]: Removed `unzip`

File: pkgs/development/compilers/flutter/default.nix [hunk 8]
```diff
@@ -21,12 +21,34 @@ let
...
+              url = "https://storage.googleapis.com/dart-archive/channels/stable/release/${dartVersion}/sdk/dartsdk-macos-arm64-release.zip";
+              sha256 = dartHash.aarch64-darwin;
+            };
           };
         };
-        src = fetchzip {
-          url = "https://storage.googleapis.com/flutter_infra_release/releases/stable/linux/flutter_linux_${version}-stable.tar.xz";
-          sha256 = hash;
-        };
+        src = rec {
```
[the remark this answers is not in the document]
Comment (devholic) [reply pkgs/development/compilers/flutter/default.nix]: Removed `rec` and replace `fetchurl` with `fetchzip` 🙏

**Comment by devholic** [comment]
@FlafyDev @hacker1024 I reverted darwin-flavored patches (which are in the base PR) and seems working as expected. Could you take a look?

**Review by FlafyDev — CHANGES_REQUESTED** [review]
I still can't test it, but overall it looks good!

left some comments

File: pkgs/development/compilers/flutter/engine-artifacts/default.nix [hunk 9]
```diff
@@ -84,19 +135,31 @@ let
   mkArtifactDerivation = { platform ? null, variant ? null, subdirectory ? null, archive, ... }@args:
     let
       artifactDirectory = if platform == null then null else "${platform}${lib.optionalString (variant != null) "-${variant}"}";
-      archiveBasename = lib.removeSuffix ".${(lib.last (lib.splitString "." archive))}" archive;
+      archiveExtension = ".${(lib.last (lib.splitString "." archive))}";
+      archiveBasename = lib.removeSuffix archiveExtension archive;
+      overrideUnpackCmd = builtins.elem archive [ "FlutterEmbedder.framework.zip" "FlutterMacOS.framework.zip" ];
```
[the remark this answers is not in the document]
Comment (FlafyDev) [reply pkgs/development/compilers/flutter/engine-artifacts/default.nix:140]: apparently [FlutterMacOS.framework.zip](https://storage.googleapis.com/flutter_infra_release/flutter/45f6e009110df4f34ec2cf99f63cf73b71b7a420/darwin-x64-release/FlutterMacOS.framework.zip) comes as a zip inside a zip.

And [FlutterEmbedder.framework.zip](https://storage.googleapis.com/flutter_infra_release/flutter/45f6e009110df4f34ec2cf99f63cf73b71b7a420/darwin-x64/FlutterEmbedder.framework.zip) fails with `replace FlutterEmbedder.framework/Resources?`

**Review by FlafyDev — APPROVED** [review]

**Comment by thiagokokada** [comment]
Result of `nixpkgs-review pr 210067` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>13 packages built:</summary>
  <ul>
    <li>firmware-updater</li>
    <li>firmware-updater.debug</li>
    <li>fluffychat</li>
    <li>fluffychat.debug</li>
    <li>flutter</li>
    <li>flutter-unwrapped (flutterPackages.stable)</li>
    <li>flutter-unwrapped.cache (flutterPackages.stable.cache)</li>
    <li>flutter37</li>
    <li>flutter37-unwrapped (flutterPackages.v37)</li>
    <li>flutter37-unwrapped.cache (flutterPackages.v37.cache)</li>
    <li>hover</li>
    <li>yubioath-flutter</li>
    <li>yubioath-flutter.debug</li>
  </ul>
</details>

**Review by thiagokokada — APPROVED** [review]

**Comment by rober-m** [comment]
Thank you! 💯

**Pull request merged**